src/cmd/ksh93/edit/edit.c

```diff
@@ -488,8 +488,7 @@ void ed_setup(Edit_t *ep, int fd, int reedit) {
                     int skip = 0;
                     ep->e_crlf = 0;
                     if (pp < ppmax) *pp++ = c;
-                    for (n = 1; *last; n++) {
-                        c = *last++;
+                    for (n = 1; c = *last++; n++) {
```

Comment:
The problem is the structure of this loop which necessitates the `if (c == 0 || c == ESC || c == '\r') last--;` that follows it. Notice that the only reason it has to decrement `last` when the end of the string is reached is because of how the code is structured. And the other two conditions in that `if` should be handled in the block containing the `break`.  Better would be to defer the increment of `last` till the end of the loop just before the next iteration.

Reply:
Also, if we decide to merge this simple fix rather than my more complex fix in PR #774, we should at least enclose the assignment in parens to eliminate the lint warning.